wolfSSL pkg addition with examples

kaleb-himes opened the pull request:
This PR will introduce wolfSSL as a pkg and contains 4 examples.

Example one is a benchmark example that will time the wolfCrypt library algorithm performance

Example two is a crypto test that ensures the wolfCrypt library is performing as expected on the device

Example three and four are a client/server example that run on 127.0.0.1 port 11111 by default and display a successful SSL/TLS connection when run from two separate terminals. Ideally either of these may be modified to use custom IP and port for testing on non-native devices and to use the available TCP/IP stack.

For any feedback/concerns/questions feel free to post on this PR or contact me directly kaleb@wolfssl.com

MichelRottleuthner commented:
Just tried it, but all examples fail to compile with `bin/pkg/native/wolfssl/wolfcrypt/src/wc_port.c:329:16: error: old-style function definition [-Werror=old-style-definition] wolfSSL_Mutex* wc_InitAndAllocMutex()`.

Once this is fixed, `wolfssl_client `and `wolfssl_server` compile and are working fine on native.

`wolfssl_crypto_benchmark` gives the following output and crashes with memory access error after the last line:

```
wolfSSL Crypto Benchmarks!
wolfCrypt Benchmark (min 1.0 sec each)
RNG      11050 kB took 1.001 seconds,   10.777 MB/s
AES-Enc  24000 kB took 1.000 seconds,   23.430 MB/s
AES-Dec  24150 kB took 1.000 seconds,   23.580 MB/s
ARC4     68225 kB took 1.000 seconds,   66.622 MB/s
RABBIT   87225 kB took 1.000 seconds,   85.162 MB/s
3DES      7925 kB took 1.001 seconds,    7.731 MB/s
MD5      118975 kB took 1.000 seconds,  116.186 MB/s
SHA      57450 kB took 1.000 seconds,   56.094 MB/s
SHA-256  24075 kB took 1.000 seconds,   23.503 MB/s
```

`wolfssl_crypto_test` still fails to compile with various old-style function definition errors in `bin/pkg/native/wolfssl/wolfcrypt/test/test.c`. After fixing them the test gives the following output and then crashes with mem access error again.

```
wolfSSL Crypto Test!
error    test passed!
base64   test passed!
base64   test passed!
MD5      test passed!
MD4      test passed!
SHA      test passed!
SHA-256  test passed!
Hash     test passed!
HMAC-MD5 test passed!
HMAC-SHA test passed!
HMAC-SHA256 test passed!
ARC4     test passed!
HC-128   test passed!
Rabbit   test passed!
DES      test passed!
DES3     test passed!
AES      test passed!
AES192   test passed!
AES256   test passed!
RANDOM   test passed!
```

Compiler version used: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609
Trying to compile it for any other platform fails for all of the examples.

kaleb-himes commented:
Thanks @MichelRottleuthner for feedback. All tested on Mac OS X. Will work on resolving the noted issues for Ubuntu 16.04.4 w/ gcc v 5.4.0

Warmest Regards,

Kaleb

kaleb-himes commented:
Hi @MichelRottleuthner,

I have pushed fixes for the old-style function defs here: https://github.com/wolfSSL/wolfssl/pull/1029

Thanks again for your time and review. The reason for the seg faults were the default stack sizes set for the application on Ubuntu. Bumping default stack size to 24.5k seems sufficient in my testing. I tested the following:

60K - Pass
30K - Pass
15K - SEGV
22.5K - SEGV
26.25 - Pass
24K - SEGV
24.5K - Pass

I did not optimize further after that.

Warm Regards,

Kaleb

kaleb-himes commented on pkg/wolfssl/Makefile:
```diff
@@ -0,0 +1,21 @@
+PKG_NAME=wolfssl
+PKG_URL=https://github.com/wolfssl/wolfssl.git
+#PKG_VERSION=a79f9c93c98f1af4e6b2c70abe8208bab8919944
+PKG_VERSION=f7cd8a0f15774f8fdb686c13008ec48601b64b28
```
Will update once PR to address old-style function defs is merged into wolfSSL master:
https://github.com/wolfSSL/wolfssl/pull/1029

MichelRottleuthner commented:
Thanks for your quick fix! Is there anything that can be done to further decrease memory usage of the test Applications to allow testing on very constrained devices?

MichelRottleuthner commented on pkg/wolfssl/Makefile:
```diff
@@ -0,0 +1,21 @@
+export RIOT_CFLAGS = ${CFLAGS} ${INCLUDES}
+
+all: git-download
+	mkdir $(PKG_BUILDDIR)/../../../native/wolfssl/
```
This and the following lines shouldn't be set to `native` statically and use `$(BOARD)` instead to allow building for other platforms.

kaleb-himes commented:
@MichelRottleuthner 

> Is there anything that can be done to further decrease memory usage of the test Applications to allow testing on very constrained devices?

Yes of course. I can work on a more optimized build this coming week. What would be an "Ideal footprint" in your mind?

> This and the following lines shouldn't be set to native statically and use $(BOARD) instead to allow building for other platforms.

Thanks I'll begin testing this on Monday!

Warm Regards,

Kaleb

kaleb-himes commented:
RIOT/pkg/wolfssl/Makefile now references merged PR with old-style function def fixes here: https://github.com/wolfSSL/wolfssl/commit/cc4ca6a44aebcbbb7f7073e963d95198fd931465

Adjusted default stack size to work for both MAC OS X (Sierra) and Ubuntu 16.04, increased from 26.25k (Ubuntu) to 35K (For MAC). Clang on mac results in less optimized binary and requires more stack.

Have a feeling this may be a never ending battle adjusting the stack on a per-device basis. SOLUTION: have it work for Native on both Mac and Linux, document stack reqs for the default examples and let developers adjust as needed for target board.

kaleb-himes commented:
@MichelRottleuthner 

Hi Michel, got busy there and now able to return to testing the suggestion. When changing Makefile from ```native``` to ```$(BOARD)``` I get the following error:

```Makefile:4: *** Recursive variable `BOARD' references itself (eventually).  Stop.```

How can I avoid this setup in the RIOT makefile system?


Regards,

Kaleb

MichelRottleuthner commented:
When you replace `native` in one of the examples [Makefiles](https://github.com/kaleb-himes/RIOT/blob/wolfSSL-PKG-ALPHA/examples/wolfssl_crypto_test/Makefile#L4) it's clear that you get that error, because that's where you define the BOARD variable initially. What I referred to in my comment are the lines below [`all: git-download`](https://github.com/kaleb-himes/RIOT/blob/wolfSSL-PKG-ALPHA/pkg/wolfssl/Makefile#L13) in the package Makefile where you use `native` in the path. This shouldn't be like that because you get another build path if you build for another platform (i.e. call `make BOARD=some-board`).

kaleb-himes commented:
Hi @MichelRottleuthner,

Perfect, thank you for clarifying!

- KH

kaleb-himes commented:
@MichelRottleuthner 

The outdated diff resulted in confusion on my end. Apologies. Fix is in.

Still have it on my list to do an "optimized" build, do you have a recommended target footprint (compiled size) and stack use in mind for this "optimized build"?

- KH

MichelRottleuthner commented:
I don't actually have a specific number in mind. I just asked out of interest because RIOT supports multiple platforms that have way smaller RAM than what is used as stack size in the wolfssl tests. Of course not all of them are realistic targets e.g. [Arduino Uno](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-Arduino-Uno) seems to be out of reach with it's 2k RAM ;) ...I'd just be interested in "how tiny is possible" to support as many RIOT platforms as possible. 
Did you already test compiling for some other platforms supported by RIOT?

kaleb-himes commented:
@MichelRottleuthner 

I have not yet had a chance to build for any of the other platforms but it is definitely on my TODO list!

> I'd just be interested in "how tiny is possible" to support as many RIOT platforms as possible.

Cool. I will do two additional examples we have a "Lean-TLS" build that I will setup and example for and also a "Lean-PSK" build that is even smaller. I'll try to get examples of each up here soon. Should I do those examples in this same PR or open a separate PR for those?

Regards,

KH

MichelRottleuthner commented:
I wouldn't mind having those examples added to this PR, as it would allow to possibly find more bugs and make the build of wolfssl more stable. Also its easier for testing in this stage. Another thing to consider is that some of the examples (e.g. the crypto test) may be better placed under tests/. I'd really like to hear other peoples opinions on this.

kaleb-himes commented:
@MichelRottleuthner 

Hi Michel,

The end of the year here has picked up in activity quite a bit. Could we schedule this for review as is with the requested changes already implemented?

The "Optimized" example is still on our TO-DO list but I am not sure we will have time to implement it before RIOT SUMMIT 2017. The wolfSSL team is eager to see this initial kick-off merged for the relationship between wolfSSL and RIOT.

- KH

MichelRottleuthner commented:
For me the optimized example is not really important and I think that can be done later. What is important: it should be possible to build and run the examples/tests at least for the hardware plattforms that have sufficient memory. If this package only works with native it doesn't make sence to add it. The last time I tried I couldn't build for any board without running into various errors. Unfortunately I won't have much time in the next weeks so it would be nice if someone else can join. @tobhe I know you are eager to use wolfSSL on RIOT ;) @PeterKietzmann we talked about this recently, do you have someone in mind who would like to help here?

tobhe commented:
@MichelRottleuthner: I am indeed eager to use this, but I also don't have much time at the moment. 
@kaleb-himes: I tried building the four examples:
On native, the test and benchmark examples are building and working, wolfssl_client and wolfssl_server both fail:
```
wolfssl_server/main.c:116:9: error: implicit declaration of function ‘wolfSSL_CTX_use_certificate_file’; did you mean ‘wolfSSL_CTX_use_certificate_buffer’? [-Werror=implicit-function-declaration]
     if (wolfSSL_CTX_use_certificate_file(ctx, CERT_FILE, SSL_FILETYPE_PEM)

wolfssl_server/main.c:124:9: error: implicit declaration of function ‘wolfSSL_CTX_use_PrivateKey_file’; did you mean ‘wolfSSL_CTX_use_PrivateKey_buffer’? [-Werror=implicit-function-declaration]
     if (wolfSSL_CTX_use_PrivateKey_file(ctx, KEY_FILE, SSL_FILETYPE_PEM)

wolfssl_client/main.c:111:9: error: implicit declaration of function ‘wolfSSL_CTX_load_verify_locations’; did you mean ‘wolfSSL_CTX_load_verify_buffer’? [-Werror=implicit-function-declaration]
     if (wolfSSL_CTX_load_verify_locations(ctx, CERT_FILE, NULL)
```
Other boards than native fail with:
```
wolfssl/wolfssl/wolfcrypt/wc_port.h:121:17: error: unknown type name 'pthread_mutex_t'
         typedef pthread_mutex_t wolfSSL_Mutex;

wolfssl/wolfssl/io.h:120:22: fatal error: sys/socket.h: No such file or directory
             #include <sys/socket.h>
```
I would recommend replacing those with riot api functions before merging.

Also, i would consider moving the test and benchmark from `examples/` to `tests/`, as that's what they are more useful for.

kaleb-himes commented:
Hi @MichelRottleuthner and @tobhe,

I got busy there for a few weeks but will be back on this in the coming week. We're eager to get this PR in before the RIOT Summit 2017.

I can make the change to move test and benchmark to "test" directory and adjust the defines to exclude threading support. In reality though there is no magic build for every platform. This is to get people access to the library to begin using it but they will need to adjust some of the settings depending on the platform. We have a porting guide for this purpose on our website here: https://wolfssl.com/wolfSSL/Docs-wolfssl-porting-guide.html

Regards,

- KH

kaleb-himes commented:
Hi @MichelRottleuthner  and @tobhe,

As requested move the tests to "tests" directory.
Did try a few builds for other platforms. Since we are an SSL/TLS implementation we rely on third-party TCP stack to be available. The only one I could find in pkgs was lwip.

Unfortunately the repo was incorrect so I updated it to the proper URL.

I am not sure who maintains that package but further errors were encountered with:
```/Users/khimes/work/RIOT-OS/RIOT/sys/auto_init/auto_init.c:145: undefined reference to `lwip_bootstrap'```

So I can concur that this will not build out-of-the-box for platforms without an existing TCP/IP stack. If we want SSL/TLS to build out-of-the-box could you point me to a current/up-to-date TCP stack supported by RIOT and how to integrate that?

Leaving the default as native for now and I have addressed the threading issue and file system build issues on non-native platforms by adding default configure options for ```SINGLE_THREADED``` and ```NO_FILESYSTEM```.

Regards,

Kaleb

kaleb-himes commented:
Hi @MichelRottleuthner and @tobhe,

I worked around the I/O issue for now by doing a build of just the crypto libraries. I have tested that build for both the SAMD21 and stm32f4discovery devices. Compiles fine, I have not actually flashed an run on either device yet.

The client and server have also been adjusted to use buffers and not rely on a filesystem as many embedded systems do not have file systems however the I/O with the TCP stack remains an issue on non-native devices at present. Looking forward to hearing from you on TCP stack availability on some of these devices.

- Kaleb

MichelRottleuthner commented:
As mentioned before I don't have much time to work on this at the moment. So I just gave it a quick try. 
I tried the "crypto-only" tests. For wolfssl_crypto_test I got this:

**samd21-xpro**: `ld: region 'ram' overflowed by 3624 bytes`
**saml21-xpro**: `ld: region 'ram' overflowed by 3600 bytes`
**samr21-xpro**: `ld: region 'ram' overflowed by 3612 bytes`

**stm32f4discovery**:
```
wolfSSL Crypto Test!
error    test passed!
base64   test passed!
base64   test passed!
MD5      test passed!
MD4      test passed!
SHA      test passed!
SHA-256  test passed!
Hash     test passed!
HMAC-MD5 test passed!
HMAC-SHA test passed!
HMAC-SHA256 test passed!
ARC4     test passed!
HC-128   test passed!
Rabbit   test passed!
DES      test passed!
DES3     test passed!
AES      test passed!
AES192   test passed!
AES256   test passed!
RANDOM   test failed!
 error = -5004
 #! exit -1: powering off
```
For wolfssl_crypto_benchmark I got the following for all of the boards above:
```
wolfssl/wolfcrypt/benchmark/benchmark.c: In function 'current_time':
wolfssl/wolfcrypt/benchmark/benchmark.c:3754:9: error: implicit declaration of function 'gettimeofday' [-Werror=implicit-function-declaration]
         gettimeofday(&tv, 0);
```

If the failing part is removed it builds. But running it on samr21-xpro and stm32f4discovery shows:
```
wolfSSL Crypto Benchmarks!
wolfCrypt Benchmark (block bytes 1048576, min  sec each)
Benchmark block buffer alloc failed!
#! exit 1: powering off
```
Regarding the TCP stack, why not use our own gnrc stack instead of lwip? Maybe @miri64 can help/point you to some examples regarding integration?

tobhe commented:
I would also recommend using GNRC's tcp implementation, which is RIOT's native tcp implementation.
For reference here is the documentation for RIOT's tcp socket implementation: [tcp-sock](http://riot-os.org/api/sock_2tcp_8h.html)

Also, native is not working anymore. Compiling with `BOARD=native` gives me: 
```
riot-wolfssl/tests/wolfssl_crypto_test/bin/pkg/native/wolfssl/src/internal.c:24140:0: error: ISO C forbids an empty translation unit [-Werror=pedantic]
 #endif /* WOLFCRYPT_ONLY */
```

rfuentess commented:
> I would also recommend using GNRC's tcp implementation, which is RIOT's native tcp implementation.
For reference here is the documentation for RIOT's tcp socket implementation: tcp-sock

Sadly, the sock_tcp is still in an experimental state as some structures [are not defined](http://riot-os.org/api/group__net__sock__tcp.html#ga0144778bc074e291fcb0d43bd35c0179). Of course, GNRC supports [TCP](http://riot-os.org/api/group__net__gnrc__tcp.html).

kaleb-himes commented:
Hi @tobhe and others,

Yes the empty translation unit error is due to inclusion of source files that contain no code source with certain settings. As I noted previously, there is no one perfect setup that will support all devices as wolfSSL has many macro definitions depending on platform. The intent of this PR was to expose the community to wolfSSL from a generic perspective. It is expected that there will be some customizations necessary on a device basis.

What do you need to see in order to get something merged in prior to the SUMMIT next week. 

examples/wolfssl_client - Running on which platform
examples/wolfssl_server - Running on which platform
tests/wolfssl_crypto_test - Running on which platform
tests/wolfssl_crypto_benchmark - Running on which platform

Or would you prefer if I removed the client/server or the tests for now, would it be best to focus on one or the other, tests or examples?

Let me know what it is you need to see specifically. Also when does the release cycle start? When do I need to have my final commit in by if this is to make it into the release?


Warm Regards,

Kaleb

tobhe commented:
@kaleb-himes  As I am not in charge to decide what is merged an what is not, I have to pass this one on to @MichelRottleuthner.
However, here are some things that I personally would like to see before a merge:

- All examples should be working on native + boards of your choice
- The examples should use RIOT's native GNRC IP/TCP stack (see @rfuentess comment)

Future improvements which I don't necessarily see mandatory for this merge:

- Improvement of the board specific config handling (this could be done in RIOT's board config directory)

kaleb-himes commented:
@tobhe,

Great! Thanks for your feedback. I will queue up the changes by Friday evening if I have the luxury of that time. Do you know when the next release cycle begins?

- Kaleb

tobhe commented:
I'm sorry, I actually have no idea when the next cycle starts.

rfuentess commented:
> All examples should be working on native + boards of your choice

Maybe @kaleb-himes  can focus in the `iotlab-m3` boards as a starting point? Or any other boards with more RAM than those?

tobhe commented:
@rfuentess This is actually a good idea
@kaleb-himes I would actually be interested in which exact parts of wolfssl need board specific configuration and/or functions

kaleb-himes commented:
@tobhe 

As noted in our porting guide here: https://wolfssl.com/wolfSSL/Docs-wolfssl-porting-guide.html

The following can all be affected by architecture, and platform:

```
2.1  Data Types
    Setting the correct data type size for your platform is always important.

2.2  Endianness
    Necessary if your platform is a big endian system.

2.3  writev
    Necessary if <sys/uio.h> is not available.

2.4  Input / Output
    Necessary if a BSD-style socket API is not available, you are using a custom transport layer or TCP/IP stack, or only want to use static buffers.

2.5  Filesystem
    Necessary if file system is not available, standard file system functions are not available, or you have a custom file system.

2.6  Threading
    Necessary if you want to use wolfSSL in a multithreaded environment, or want to just compile it in single threaded mode.

2.7  Random Seed
    Necessary if either /dev/random or /dev/urandom is not available or you want to integrate into a hardware RNG.

2.8  Memory
    Necessary when you don’t have standard memory functions available or are interested in memory usage differences between optional math libraries.

2.9  Time
    Necessary when standard time functions are not available, or you need to define a custom clock tick function.

2.10  C Standard Library
    Necessary when a C standard library is not available, or using a custom one.

2.11  Logging
    Necessary when debug messages desired but stderr is unavailable.

2.12  Public Key Operations
    Necessary if you want to use your own public key implementation.

2.13  Atomic Record Layer Processing
    Necessary if you want to do your own processing of record layers, specifically MAC/encrypt and decrypt/verify operations.

2.14  Features
    Necessary when you want to disable features for example if memory constrained you could disable features to reduce footprint size.
```

tobhe commented:
As far as I can see, there is actually not much board specific configuration that isn't solved already:
> 2.1  Data Types
    Setting the correct data type size for your platform is always important.

I'm sure this is configured for each supported board and can be found somewhere in `boards/` or `cpu/`

> 2.2  Endianness
    Necessary if your platform is a big endian system.

Same as above

> 2.3  writev
    Necessary if <sys/uio.h> is not available.

Should be available on every board via posix interface, I'm not sure if this is necessary

> 2.4  Input / Output
    Necessary if a BSD-style socket API is not available, you are using a custom transport layer or TCP/IP stack, or only want to use static buffers.

This should be configured to use GNRC if possible

>  2.5  Filesystem
    Necessary if file system is not available, standard file system functions are not available, or you have a custom file system.

Doesn't exist

> 2.6  Threading
    Necessary if you want to use wolfSSL in a multithreaded environment, or want to just compile it in single threaded mode.

This could be left out and eventually ported later if necessary

>  2.7  Random Seed
    Necessary if either /dev/random or /dev/urandom is not available or you want to integrate into a hardware RNG.

I would suggest using RIOT's random module, initialized with a hwrng if available

> 2.8  Memory
    Necessary when you don’t have standard memory functions available or are interested in memory usage differences between optional math libraries.

This could be the biggest problem as free is currently not doing anything, also fastmath

> 2.9  Time
    Necessary when standard time functions are not available, or you need to define a custom clock tick function.

Not sure if this is available or must be configured, but it should be uniform between all supported boards

> 2.10  C Standard Library
    Necessary when a C standard library is not available, or using a custom one.

Should be working

> 2.11  Logging
    Necessary when debug messages desired but stderr is unavailable.

Uniform between boards

> 2.12  Public Key Operations
    Necessary if you want to use your own public key implementation.

We don't have one as far as I can see

> 2.13  Atomic Record Layer Processing
    Necessary if you want to do your own processing of record layers, specifically MAC/encrypt and decrypt/verify operations.

Not sure if this is needed but it's not board specific

> 2.14  Features
    Necessary when you want to disable features for example if memory constrained you could disable features to reduce footprint size.

This is mostly application specific 

So summarizing the above: if we find out where to get each boards data type sizes and endianness, we should be able to solve most of the configuration for all supported boards in a generic way.

smlng commented on pkg/lwip/Makefile:
```diff
@@ -1,5 +1,5 @@
 PKG_NAME=lwip
-PKG_URL=git://git.savannah.nongnu.org/lwip.git
+PKG_URL=git://git.savannah.nongnu.org/git/lwip.git
```
unrelated? Otherwise: please explain.

kaleb-himes replied:
Was trying to get a TCP/IP stack before someone pointed out gnrc_tcp. In attempting to use pkg lwip kept getting a failure to resolve the github repo. On investigating found the repo URL was missing the "/git" so I updated it to test with pkg lwip. I will remove that change in my next commit.

emmanuelsearch commented:
@kaleb-himes instead of a TLS example initially, it would probaly have more impact to first focus on providing an example of one could use the WolfSSL package to provide DTLS over UDP and IPv6/6LoWPAN, by extending an existing example (such as [gnrc_networking](https://github.com/RIOT-OS/RIOT/tree/master/examples/gnrc_networking) for instance). What do you think about this?

smlng commented:
@kaleb-himes how to proceed with this?

smlng commented:
this needs work, but should stay on release schedule -> postpone

kaleb-himes commented:
@smlng,

Sorry for the delay, yes this needs work but I agree would like to keep it on release schedule -> postpone. As the full time support guy here at wolfSSL I have been extremely busy without time to dedicate to testing and implementing support for specific target devices. I would like to add support for at least an Atmel (Now microchip) and one or two other devices before we target a merge.

kaleb-himes commented:
Quick question (since IRC is pretty dead on the weekends)
Trying to test our crypto sources on a samd21 xplained pro and seeing major optimization failures by the compiler. Am I missing a setting somewhere? Do I need to install a different toolchain?

wolfcrypt_test/bin/native/wolfssl/wolfcrypt/src

```
-rw-r--r--   1 kalebhimes  staff   8.7K Jan 28 16:42 sha256.d
-rw-r--r--   1 kalebhimes  staff   8.0K Jan 28 16:42 sha256.o
```

Update makefile line from ```BOARD ?= native``` to ```BOARD ?= samd21-xpro```
wolfcrypt_test/bin/samd21-xpro/wolfssl/wolfcrypt/src

```
-rw-r--r--   1 kalebhimes  staff   5.9K Jan 28 16:46 sha256.d
-rw-r--r--   1 kalebhimes  staff    87K Jan 28 16:46 sha256.o. <--- why so much bloat??
```

smlng commented:
Hi @kaleb-himes, (IIRC) on native the object file does not contain external shared libraries it depends on, but for samr21-xpro the target is cross-compiled, hence the object file contains all external library function. You should compare the binaries sizes the linker reports at the end of `make`, but even then you'll see (large) differences between varying platforms, e.g., thread stack sizes differ by cpu and specifically for native they are rather large.

Maybe @gebart and @kaspar030 can also give some more info and reasoning on this?

kaleb-himes commented:
Hi @smlng,

Even if I compile that file on a Mac osx with full debug symbols the largest I can get sha256 to compile is. I still think there is something else going on here. Any thoughts?

```
21K Jan 25 17:16 src_libwolfssl_la-sha256.o
```

gebart commented:
@kaleb-himes did you look into where the bytes are located? Useful tools include size, objdump and nm

kaleb-himes commented:
@gebart,

Thank you so much for the suggestions! The major factor apparently contributing to the bloat are listed below, I've included line numbers for reference and I am attaching the result of running ```objdump -D aes.o```. It is quite obviously debug sections causing the issues. I have already set:

```
CFLAGS += -NDEBUG
```

Apparently that is not the correct flag to be using for disabling debugging. Could you tell me how to turn off debugging?

[objdump-report.txt](https://github.com/RIOT-OS/RIOT/files/1724573/objdump-report.txt)


```
 4128 Disassembly of section .debug_info: 
 5820 Disassembly of section .debug_abbrev: 
 6116 Disassembly of section .debug_loc: 
 6761 Disassembly of section .debug_aranges:
 6787 Disassembly of section .debug_ranges: 
 6829 Disassembly of section .debug_macro: 
... you'll get the idea below...
 8659 Disassembly of section .debug_macro: 
 8669 Disassembly of section .debug_macro:
 8682 Disassembly of section .debug_macro:
 8701 Disassembly of section .debug_macro:  
 8711 Disassembly of section .debug_macro:
 8737 Disassembly of section .debug_macro:
 8780 Disassembly of section .debug_macro:
 8809 Disassembly of section .debug_macro: 
 8825 Disassembly of section .debug_macro: 
 8838 Disassembly of section .debug_line: 
 9464 Disassembly of section .debug_str:
Goes to line
18109 in the disassembly.
```

gebart commented:
The debugging sections are only symbol definitions for gdb, they are not copied to the actual device ROM. Use `size` to figure out how much memory is actually used.
For the samr21-xpro: 
```
arm-none-eabi-size bin/samr21-xpro/my/object/path/sha256.o
```

MichelRottleuthner commented:
@kaleb-himes how is it going? I'd really like to move forward with this. 
I tried to run the client/server on real hardware, but there are still some compile errors. After resolving them by fixing a few includes and adding lwip I got it to compile but it didn't work (failed somewhere in socket implementation of lwip).
Because the "default" transport layer used on RIOT is UDP I think it would make sense to focus on a simple DTLS example (like @emmanuelsearch stated above). That way we get rid of the lwip-dependency and can use RIOTs own network stack. Adapting the wolfSSL [dtls example](https://github.com/wolfSSL/wolfssl-examples/blob/master/dtls/server-dtls.c) should be a good starting point, no? A look at the [dtls-echo example](https://github.com/RIOT-OS/RIOT/tree/master/examples/dtls-echo) could help with the necessary modules in the Makefile and maybe some functions for a shell interface of the example. 
Let me know if I can help you with anything so we can finally get this going :)

kaleb-himes commented:
@MichelRottleuthner 

That would be great! Thank you for providing a clear path forward and goal, it makes it much easier for me to allocate / dedicate the time to it. I'll see if I can free up some room in my schedule this week to start working on putting together a DTLS example that will work and test it on two separate devices!

- KH

kaleb-himes commented:
@MichelRottleuthner,

Apologies my other obligations have delayed any updates. One of my colleagues will be attending RIOT Summit 2018 in September @danielinux. I have sent him your comments and added him as a collaborator on this PR! Many thanks for your patience as we have many tasks competing for priorities RIOT OS is right at the top of our list for open source efforts and we very much look forward to continued efforts between wolfSSL and the RIOT Community!

- Kaleb

mfrey commented:
@danielinux so we can close this issue? do you mind updating https://www.wolfssl.com/wolfssl-riot-os-examples/ to the latest PR? i've started there and ended up in #9894
